hazelcast/src/main/java/com/hazelcast/spi/impl/waitnotifyservice/impl/WaitNotifyServiceImpl.java

```diff
@@ -308,16 +310,35 @@ private boolean doRun() throws Exception {
             }
 
             for (Queue<WaitingOperation> q : mapWaitingOps.values()) {
-                for (WaitingOperation waitingOp : q) {
+                while (it.hasNext()) {
                     if (Thread.interrupted()) {
                         return true;
                     }
+                    WaitingOperation waitingOp = it.next();
                     if (waitingOp.isValid() && waitingOp.needsInvalidation()) {
-                        invalidate(waitingOp);
+                        Address targetAddress = getTargetAddress(waitingOp);
+                        if (!waitingOp.validatesTarget() || nodeEngine.getThisAddress().equals(targetAddress)) {
+                            invalidate(waitingOp);    
+                        } else {
+                            logger.warning("Removing operation " + waitingOp +
+                                " from queue because the local node is not the target and returning exception to invoker");
+                            WrongTargetException wte = new WrongTargetException(nodeEngine.getThisAddress(), targetAddress,
+                                waitingOp.getPartitionId(), waitingOp.getReplicaIndex(),
+                                waitingOp.getClass().getName(), waitingOp.getServiceName());
+                            OperationResponseHandler responseHandler = waitingOp.getOperation().getOperationResponseHandler();
+                            responseHandler.sendResponse(waitingOp.getOperation(), wte);
+                            it.remove();
+                        }
                     }
                 }
             }
             return false;
         }
     }
+
+    private Address getTargetAddress(WaitingOperation waitingOp) {
+        int partitionId = waitingOp.getPartitionId();
+        InternalPartition internalPartition = nodeEngine.getPartitionService().getPartition(partitionId);
+        return internalPartition.getReplicaAddress(waitingOp.getReplicaIndex());
```

Comment:
this part is tricky: this will give you an address of a primary replica. (=the partition owner). 

However in theory an operation can be targeted to a non-primary partition (=backup). I *think* for now the backup operations are never blocking, but in theory there is nothing preventing blocking backup operations.

You might want to have a look at `Operation::getReplicaIndex`.